File: js/3-Longest-Substring-Without-Repeating-Characters.js
/**

 * @param {string} s

 * @return {number}

*/

var lengthOfLongestSubstring = function(s) {
    var res=0, tmp=0,cur=0, head=0,tail=0;
    var hash=[]
    while (tail < s.length) {
        if(typeof hash[s[tail]] == 'number'){
            for(;head<=hash[s[tail]];head++){
                hash[s[head]]=null;
                tmp--;
            }
        }
        ++tmp;
        hash[s[tail]]=tail++;
        if (tmp > res) {
            res = tmp;
        }
    }
    return res;
}

File: python/3-Longest-Substring-Without-Repeating-Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        head, tail, res = [0] *3 
        hash = [0] * 200
        while tail < len(s):
            v = ord(s[tail])
            while hash[v] and head < hash[v]:
                hash[ord(s[head])] = 0
                head += 1
            tail += 1
            hash[v] = tail
            if tail-head > res : res = tail - head 
        return res 

                

a = Solution()
print a.lengthOfLongestSubstring('')
print a.lengthOfLongestSubstring('aaa')
print a.lengthOfLongestSubstring('babac')
